internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,111 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the Token method.
+type Tokener interface {
+	Token(context.Context) (string, error)
+}
+
+// LazyToken is a struct that handles token acquisition and caching.
+type LazyToken struct {
+	// Url represents the URL used for acquiring the token.
+	Url string
+	// ClientId is the client ID used for authentication.
+	ClientId string
+	// ClientSecret is the client secret used for authentication.
+	ClientSecret string
+	// token holds the currently cached token.
+	token string
```

Comment:
Please export the `Token` as well, you can use it as pre-configured (offline) token if needed.